File: posixlib/src/main/resources/scala-native/dirent.c
#if defined(SCALANATIVE_COMPILE_ALWAYS) || defined(__SCALANATIVE_POSIX_DIRENT)
#if defined(__unix__) || defined(__unix) || defined(unix) ||                   \
    (defined(__APPLE__) && defined(__MACH__))
#include <dirent.h>
#include <string.h>
#include <stdio.h>
#include <errno.h>

// Check consistency of Scala Native and operating system definitions.

_Static_assert(sizeof(ino_t) <= 64,
               "os ino_t does not match Scala Native dirent");

_Static_assert(sizeof(((struct dirent *)0)->d_type) <= 2,
               "os dirent.d_type does not match Scala Native dirent");

#ifdef NAME_MAX
/* TL;DR - The Scala Native d_name definition should be changed to CString
 *         as soon as breaking changes are allowed. That punts allocation
 *         of storage with sufficient size for a file name to the
 *         operating system, which knows best.
 *
 *         Scala Native can always to a strlen/strnlen of the returned
 *         string and allocate, at its level, the exact required size.
 *         Scala Native code while almost always be copying the os d_name
 *         to a Scala 'String'. fromCString(os_d_name) can be done
 *         directly, and does not require an intermediate allocation and
 *         copy
 *
 * In 2024 and possibly much earlier, linux & FreeBSD  "getconf NAME_MAX /"
 * return 255 for the file systems the most commonly used on those operating
 * systems. They may or may not define NAME_MAX and readdir() may or
 * may not return strings longer that that; long story.
 *
 * macOS limits need to be explored. It specifies a max of 255 UTF-8
 * characters. Since each UTF-8 character can be up to 5 bytes long,
 * a 255 UTF-8 characters could be (255 * 5 == 1275) bytes.
 *
 * Then again, there are file systems, such as Amazon S3 which has a
 * file name length limit of 1024 characters)
 *
 */
_Static_assert(NAME_MAX <= 255, "NAME_MAX does not match Scala Native dirent");
#else
#define NAME_MAX 255 // manually match dirent.scala definition
#endif

struct scalanative_dirent {
    unsigned long long d_ino;  /** file serial number */
    char d_name[NAME_MAX + 1]; /** name of entry */
    short d_type;
};

int scalanative_dt_unknown() { return DT_UNKNOWN; }
int scalanative_dt_fifo() { return DT_FIFO; }
int scalanative_dt_chr() { return DT_CHR; }
int scalanative_dt_dir() { return DT_DIR; }
int scalanative_dt_blk() { return DT_BLK; }
int scalanative_dt_reg() { return DT_REG; }
int scalanative_dt_lnk() { return DT_LNK; }
int scalanative_dt_sock() { return DT_SOCK; }
int scalanative_dt_wht() {
#ifdef DW_WHT
    return DT_WHT;
#else
    return 0;
#endif
}

DIR *scalanative_opendir(const char *name) { return opendir(name); }

void scalanative_dirent_init(struct dirent *dirent,
                             struct scalanative_dirent *my_dirent) {
    my_dirent->d_ino = dirent->d_ino;

    // Note: code will _silently_ truncate any long d_name returned by OS.
    strncpy(my_dirent->d_name, dirent->d_name, NAME_MAX);
    my_dirent->d_name[NAME_MAX] = '\0';
    my_dirent->d_type = dirent->d_type;
}

// This function is for Scala Native internal use only.

// returns 0 in case of success, -1 in case of empty dir, errno otherwise
int scalanative_readdirImpl(DIR *dirp, struct scalanative_dirent *buf) {
    errno = 0;
    struct dirent *orig_buf = readdir(dirp);
    int result = 0;

    if (orig_buf != NULL)
        scalanative_dirent_init(orig_buf, buf);
    else if (errno == 0)
        result = -1;
    else
        result = errno;

    return result;
}
/* This function is for Scala Native internal use only.
 * It is provided to preserve the entry point and prevent a breaking
 * change in 0.5.2. To reduce enduring & compounding complexity, it should
 * be removed for 0.6.0.
 *
 * No Scala Native code defines or calls it.
 */

int scalanative_readdir(DIR *dirp, struct scalanative_dirent *buf) {
    scalanative_readdirImpl(dirp, buf);
}

int scalanative_closedir(DIR *dirp) { return closedir(dirp); }

#endif // Unix or Mac OS
#endif

File: posixlib/src/main/scala/scala/scalanative/posix/dirent.scala
package scala.scalanative
package posix

import scalanative.posix.sys.types.ino_t

import scala.scalanative.unsafe._, Nat._

private[scalanative] object DirentTypes {
  type _256 = Digit3[_2, _5, _6] // see comment above 'type dirent' below.

  type DIR = CStruct0 // An opaque structure from os. Deconstruct a your peril.

  /* Use "glue" is here because Direct call through to the operating
   * system is beyond the scope of time available. Linux, FreeBSD and,
   * macOS appear to use different layouts, possibly containing padding.
   */

  /* The second field, d_name, is suspect and should be changed to CString
   * as soon as breaking changes are allowed after 0.5.0.
   *
   * The handling of d_name across operating systems is complex. All, so far,
   * agree that the field is a null-terminated array of characters, a.k.a
   * CString.
   *
   * Guarantees of the allocated size of that CString are stated but
   * unreliable.  Some operating systems do not specify NAME_MAX.
   * Some specify it as 255 but can return a string longer that that.
   * macOS specifies 255 UTF-8 characters. Since each UTF-8 character
   * can be up to, currently, 5 bytes long, the total size can
   * exceed the _256 here.
   *
   * Using a CString here, follows the Open Group 'char   d_name[]'.
   * By C rules, that is equivalent to 'char *', a.k.a CString.
   */

  /* Code in dirent.c manually checks that the sizes & declarations
   * of this structure are compatible with the operating system definitions.
   * Devo: if you change anything here, please examine the dirent.c checks.
   *
   * Re: lack of ino_t.
   *     One would expect to see the _1 field below declared as ino_t.
   *     The CUnsignedLongLong is surprising but should not be changed.
   *     dirent.c defines its idea of a Scala Native dirent using "long long".
   *     That gives 64 bytes on both 32 & 64 bit systems.
   *
   *     Scala Native defines 'ino_t' as CUnsignedLong, which is 32 bits
   *     on a 32 bit system.  Using CUnsignedLongLong (guaranteed 64 bits)
   *     allocates space sufficient for  both the 32 & 64 bit cases.
   *     It also is robust to dirent.c being compiled with
   *     _FILE_OFFSET_BITS=64 on 32 bit systems, perhaps by user specified
   *     compilation flags.
   */

  type dirent =
    CStruct3[CUnsignedLongLong, CArray[CChar, _256], CShort]
}

// not __SCALANATIVE_POSIX_DIRENTIMPL for historical/version_compatibility.
@define("__SCALANATIVE_POSIX_DIRENT")
@extern
private[scalanative] object DirentImpl {
  import DirentTypes._

  type DIR = DirentTypes.DIR

  def scalanative_closedir(dirp: Ptr[DIR]): CInt = extern

  def scalanative_opendir(name: CString): Ptr[DIR] = extern

  def scalanative_readdir(dirp: Ptr[DIR]): CInt = extern

  // Not POSIX, javalib & scalanative/nio/fs/ internal use only.
  def scalanative_readdirImpl(dirp: Ptr[DIR], buf: Ptr[dirent]): CInt = extern

  def scalanative_dt_unknown: CInt = extern
  def scalanative_dt_fifo: CInt = extern
  def scalanative_dt_chr: CInt = extern
  def scalanative_dt_dir: CInt = extern
  def scalanative_dt_blk: CInt = extern
  def scalanative_dt_reg: CInt = extern
  def scalanative_dt_lnk: CInt = extern
  def scalanative_dt_sock: CInt = extern
  def scalanative_dt_wht: CInt = extern
}

object dirent {
  import DirentImpl._

  type DIR = DirentTypes.DIR
  type dirent = DirentTypes.dirent

  def closedir(dirp: Ptr[DIR]): CInt = scalanative_closedir(dirp)

  def opendir(name: CString): Ptr[DIR] = scalanative_opendir(name)

  def readdir(dirp: Ptr[DIR]): Ptr[dirent] = {
    val entry = new Array[Byte](sizeOf[dirent]).at(0).asInstanceOf[Ptr[dirent]]

    if (DirentImpl.scalanative_readdirImpl(dirp, entry) != 0) null
    else entry
  }

  @deprecated(
    "Not POSIX, subject to removal in the future. Use readdir(dirp) instead.",
    since = "posixlib 0.5.2"
  )
  // Not to be confused with POSIX readdir_r().
  def readdir(dirp: Ptr[DIR], buf: Ptr[dirent]): CInt =
    DirentImpl.scalanative_readdirImpl(dirp, buf)

  /* readdir_r() is specified in Open Group 2018 version, and probably earlier.
   *
   * macOS allows the function but Linux & FreeBSD describe it a "deprecated"
   * and recommend using a current thread-safe version of 'readdir()'.
   * For one discussion, see:
   *  https://man7.org/linux/man-pages/man3/readdir_r.3.html
   *
   * Not implemented to avoid spending time on complexity to no good end.
   */
  // def readdir_r(/* three arguments */)

  def DT_UNKNOWN: CInt = scalanative_dt_unknown

  def DT_FIFO: CInt = scalanative_dt_fifo

  def DT_CHR: CInt = scalanative_dt_chr

  def DT_DIR: CInt = scalanative_dt_dir

  def DT_BLK: CInt = scalanative_dt_blk

  def DT_REG: CInt = scalanative_dt_reg

  def DT_LNK: CInt = scalanative_dt_lnk

  def DT_SOCK: CInt = scalanative_dt_sock

  def DT_WHT: CInt = scalanative_dt_wht
}

object direntOps {
  implicit class direntOps(val ptr: Ptr[dirent.dirent]) extends AnyVal {
    def d_ino = ptr._1
    def d_name = ptr._2.at(0).asInstanceOf[CString]
    def d_type = ptr._3
  }
}
